### PR TITLE
Pin the VM lane's environment against daily-stable.yml's, and carry the three variables it never had

### LINKED ISSUE #1717: Nothing pins the VM lane's env against daily-stable.yml's, and the silent half of that gap cannot be found by running it
Follow-up to #1714, found while reviewing #1716.

#1714 was one variable — `LANGFLOW_DEACTIVATE_TRACING` — pinned off in the source starter while `daily-stable.yml` runs it on. It cost a full VM run to surface and one PR to fix. It is an instance of a class, and the rest of that class is still open.

## What is measured

`daily-stable.yml` sets 12 `LANGFLOW_*` variables on its service. The VM path — `scripts/run-e2e.sh` → ssh → `scripts/start-langflow-source.sh` — sets 8. Discounting the ones that cannot apply to a source instance (`LANGFLOW_IMAGE` / `LANGFLOW_VERSION` select an image; `LANGFLOW_WORKERS` the starter carries as `--workers`), **three** are absent from the whole VM path — not in `run-e2e.sh`, not in `start-langflow-source.sh`, not in `start-langflow-pip.sh`:

| Variable | What the divergence costs |
|---|---|
| `LANGFLOW_ALLOW_CUSTOM_COMPONENTS` | **Red.** The product default is `false` (`docs/core-components/full-custom-component.md:125`; `daily-stable.yml`'s own comment records that the nightly image ships `false` deliberately), so every spec declaring it as a precondition — `core-components/full-custom-component`, `security/code-execution-endpoints` and the siblings — cannot pass on the VM. Plausibly the 9th failure of the run #1714 was raised from, which the tracing fix does not address |
| `LANGFLOW_SQLITE_PRAGMAS` (`foreign_keys:ON`) | **False green.** With SQLite foreign keys off, a cascade/orphan defect (the #13955 class) is invisible on the VM and visible on Actions |
| `LANGFLOW_WORKER_TIMEOUT` | Changes wedge/timeout behaviour (the #922/#927 class), so the two lanes attribute flakes differently |

## Why this is worth a mechanism and not three lines

The first one is discoverable by running the lane: it turns red, ~50 minutes per discovery, which is the cycle #1714 just paid.

The second is **not discoverable by running the lane at all** — it produces agreement, not a failure. And this lane exists to *compare verdicts* with the Actions daily, so a green that should not be green corrupts the exact product stage 1 of the migration is built to deliver. No amount of running exposes that half of the class.

## Proposal

Fix the three, and — the point of the issue — add the fail-closed guard so the enumeration is not a one-time hand audit: every `LANGFLOW_*` in `daily-stable.yml`'s service env must either **reach the target** or be **recorded out of scope with a reason**, in the shape `scripts/watch-upstream-areas.mjs --mode=check` already uses for the `lfx` subtrees. That way the 13th variable the daily gains forces a decision when it is written, instead of opening a silent gap that surfaces whenever someone next notices an odd verdict.

Two details the guard has to get right, both learned in #1716's review:

- It reads workflow YAML, so `stripComments()` first and **every** occurrence must be defensible, never the first match only — the failure `watch-tokens.test.mjs:1276-1290` documents and the one #1716's first draft reintroduced (a prose comment naming a variable both masked a real value and failed a correct workflow).
- A value that is a `${{ … }}` expression goes through `scripts/lib/gh-expression.mjs`, not a regex — #1226's rule.

## Acceptance

- The three variables reach the instance the VM lane serves, or are classified out of scope with the reason recorded next to the classification.
- A test in the `npm run test:scripts` lane fails when `daily-stable.yml` gains a `LANGFLOW_*` the VM path neither forwards nor classifies.
- The guard is pinned by mutation, not by spelling: adding an unclassified variable to the workflow fails it, and a comment mentioning a variable does not.

### PR DESCRIPTION
**Pull request by Victor-w-Madeira**
Closes #1717.

## What was measured

`daily-stable.yml` sets **ten** `LANGFLOW_*` variables on the Langflow service it tests. The VM path — `run-e2e.sh` → ssh → `start-langflow-source.sh` — carried **seven**. Three never reached the instance, and they are not equally visible:

| Variable | What its absence costs |
|---|---|
| `LANGFLOW_SQLITE_PRAGMAS` | **The silent one.** `foreign_keys: ON` is what makes a cascade/orphan defect (upstream #13955) visible. Without it the raw DELETE "succeeds" leaving orphaned rows, so `traces-delete-cascade` passes on the VM and fails on Actions — the comparison reports agreement, and no run can refute it |
| `LANGFLOW_ALLOW_CUSTOM_COMPONENTS` | **Latent agreement.** The workflow sets it to override a default the nightly *image* bakes in; a source instance bakes in nothing, so the two lanes agree today for different reasons. The day the product default moves, only one lane notices |
| `LANGFLOW_WORKER_TIMEOUT` | Bounds what one wedge costs (#1048), so the two lanes attribute flakes the same way |

The issue's table marked `LANGFLOW_ALLOW_CUSTOM_COMPONENTS` **Red** and offered it as the ninth failure of the run that raised #1714. That was already corrected in the issue's own comments, and this PR follows the correction: measured on the machine, the custom-component specs pass **8/8** on the VM without the variable. It is mirrored because it is latent, not because it is failing.

## Where the fix lives, and why not in the starter

In `run-e2e.sh`. The starters' environment blocks are asserted identical to `start-langflow-pip.sh`'s precisely so **a spec cannot tell which starter brought its instance up**; a lane-specific value written there would either break that assertion or defeat its purpose. Same relocation #1716 made for tracing.

No starter change was needed, and the mechanism is worth stating because it is the load-bearing part: a `VAR=value` prefix on the remote `bash -s` lands in **that shell's environment**, and the starter's own settings are prefix assignments on `uv run`, which *add to* the inherited environment rather than replace it. The server sees both.

## The quoting bug the new values exposed

`ssh` joins its arguments and hands **one** string to a shell on the far side, so anything unquoted is re-split there. Every mirrored value so far was a bare word and survived that by luck. `LANGFLOW_SQLITE_PRAGMAS` is a JSON object: unquoted, the variable arrives as `{"synchronous":` with five loose words after it.

`shq` quotes for the remote shell and `mirrored_target_env` composes the prefix as a **function**, so the tests can read exactly what would be sent — quoting included — without a machine to send it to.

## The guard, which is the point of the issue

Three lines in a script fix today. The enumeration is what makes the 11th variable force a decision when it is written:

```
node scripts/check-vm-env-parity.mjs --mode=check   # fail-closed verdict
node scripts/check-vm-env-parity.mjs --mode=list    # the table
```

Every `LANGFLOW_*` on the workflow's service must either **reach the target** or be **recorded out of scope with a reason** — the shape `watch-upstream-areas.mjs --mode=check` already uses. It runs in `npm run test:scripts`, and as a CLI too, because the person holding an odd verdict is on the QA VM.

Two things the reader gets right, both learned in #1716's review:

- **full-line comments are stripped first**, and a comment naming a variable counts in *neither* direction — it must not mask a real setting, and it must not fail a correct workflow (#1300)
- **the read is scoped to the service block.** `daily-stable.yml` also sets `LANGFLOW_IMAGE` and `LANGFLOW_VERSION` as step env in its reporting jobs; counted as service variables they would be permanent, unfixable gaps. That is also why the count here is 10 and the issue said 12
- values go through `scripts/lib/gh-expression.mjs`, not a regex — #1226's rule. The first draft of the reader duplicated that unquoting and would have been a second answer able to disagree with the first

## Machine exceptions still work, deliberately

Every mirrored value stays overridable. That is how a **machine** records a measured exception — the qa VM overrides tracing while #1720 is open, with the reason beside it. What is pinned is the **default**, and it is read *out of the workflow* rather than copied, so a value changed there cannot leave this one behind. The parity test loops over the classification, so the next variable to be mirrored is covered by it on the day it is classified.

Pragma **content** is left to the operator for the same reason; only parseability is enforced, because malformed JSON is never intentional and Langflow falls back to its own defaults without saying so.

## Pinned by mutation, and two of the pins exist because the mutation failed first

- removing a variable from the composer → caught
- a workflow default changed without the script following → caught
- **the far-side test did not catch the first case.** It ran the remote shell with *this* shell's environment, so the value arrived by inheritance and the assertion stayed green with the composer emptied — the exact failure it was written for, one layer down. `env -u` per mirrored name is what makes it a real boundary
- **the first `shq` was wrong.** The `${1//\'/…}` form eats its backslashes twice inside double quotes and produced `'it\'\\'\'s'`, which does not parse. No current value carries a quote, so nothing else in the suite would have found it; there is now a test for the branch none of today's values reach
- guard side: an unclassified variable fails; a comment naming one does not; a classification claiming a carrier that never names the variable fails; a carrier's own *comment* does not count as carrying; out of scope with no reason is not out of scope; a stale entry fails; an unreadable workflow fails rather than reading as clean (#1012)

Writing the classification is what found the first gap: `LANGFLOW_A2A_ENABLED` was missing from it, and the guard said so on its first run against the real repository.

## Verification

- `npm run test:scripts` — 1116 passing, 0 failing. Each of the three commits is green on its own.
- `npm run lint` — 0 errors (pre-existing warnings only, none in the changed files).
- The end-to-end mutation: adding an unclassified `LANGFLOW_*` to the real `daily-stable.yml` fails the lane and names it.
- The composed remote environment, round-tripped through a shell the way ssh would: the pragmas arrive **parseable**, with `foreign_keys: ON`.

Not smoked on the VMs yet — the run this affects is Monday's, and it only picks the change up once the mirror has synced.

Unrelated, noticed while running the lane repeatedly: `scripts/wait-for-backend.test.mjs`'s "healthy backend: returns ok on the first attempt" flaked once in five full runs under load. Untouched here.

### CODE REVIEW

**Comment by Victor-w-Madeira** [comment]
## Smoked on the machines — the body said "not smoked yet", and now it is

Run against `langflowbin` on port 7862, through the real chain (`ssh` → `bash -s` → `start-langflow-source.sh` → `uv run`), with the environment composed by `mirrored_target_env` itself rather than retyped. Clone at `0c83e594` / `v1.13.0.dev2`, build stamp valid, ready in 10 s.

The environment of the process that actually **serves**:

```
$ tr '\0' '\n' < /proc/57352/environ | grep -E '^LANGFLOW_(DEACTIVATE|WORKER_TIMEOUT|ALLOW_CUSTOM|SQLITE)'
LANGFLOW_ALLOW_CUSTOM_COMPONENTS=true
LANGFLOW_DEACTIVATE_TRACING=false
LANGFLOW_SQLITE_PRAGMAS={"synchronous": "NORMAL", "journal_mode": "WAL", "busy_timeout": 30000, "foreign_keys": "ON"}
LANGFLOW_WORKER_TIMEOUT=120
```

and the child that gunicorn serves from carries `LANGFLOW_SQLITE_PRAGMAS` too — which is the half the unit tests cannot reach, because it depends on `uv run` passing the inherited environment through to the server rather than on anything this repo controls. `GET /api/v1/version` answered `1.13.0.dev2`. Stopped cleanly by `stop-langflow-source.sh`, no orphan, state directory removed.

**Scope of the claim, stated so nobody reads more into it:** this proves the four values *reach the serving process*, which is the mechanism this PR changes. It does not prove Langflow honours `LANGFLOW_SQLITE_PRAGMAS` — that is the product's own contract, identical code to the image the Actions lane runs, and `traces-delete-cascade.spec.ts` is what actually exercises it.

**Unrelated, found while checking the machine was clean.** PID 55299 on `langflowbin` is an orphaned Langflow from an earlier run: alive 1 h 57 m, `PPID=1`, 58 threads, answering nothing, **listening on no port**, and its state directory `/tmp/langflow-source-7860` has no `langflow.pid` — so `stop-langflow-source.sh` cannot stop it. It does not block a future run (the port is genuinely free, so both of the starter's refusal checks pass correctly), but it is exactly the shape the PID-file debt predicts: *"a PID file that guards only a number, and the number is the launcher's"*. Worth a line on the divergence list rather than a fix here.

**Review by Victor-w-Madeira — COMMENTED** [review]
## Review

The mechanism is the right one and the three values are right. Reviewed against the real sources rather than the PR body: the branch at `4e7c75b5`, `origin/main` at `d3af7349` (so #1716 is in the base), `daily-stable.yml`'s service block, both starters, and upstream `lfx` for the two product-default claims.

One change I would ask for before merge — the guard has exactly one door left open, and it is the door #1716 already walked through once. Everything else below is minor.

### Verified independently

- **10** `LANGFLOW_*` on `daily-stable.yml`'s Langflow service, and `CLASSIFICATION` covers all 10. `--mode=check` green, `--mode=list` readable, exit codes as documented.
- `DEFAULT_SQLITE_PRAGMAS` is byte-identical to the workflow's line, and the comparison is *read* through `evaluateWorkflowValue`, not copied — the single-quoted YAML scalar round-trips and `foreign_keys` comes back `ON`.
- `npm run test:scripts`: **1117 tests, 1116 pass, 0 fail**, and each of the three commits is green on its own (1017 / 1022 / 1116 passing) — as claimed.
- `shq` is correct as written: inside the double-quoted `sed` string `\\\\` collapses to `\\`, so the replacement is `'\''`. The `${1//\'/…}` note is right too — worth keeping the comment, since the correct form looks like the broken one.
- The two product-default claims that changed the issue's severity call both hold: `lfx/services/settings/groups/security.py:74` reads `allow_custom_components: bool = True`, so the workflow really is overriding an *image* default and the "latent agreement, not red" re-marking is correct; and `LANGFLOW_WORKER_TIMEOUT` does reach gunicorn on this lane — `__main__.py:549` → `options["timeout"]` at `:684`, and Linux takes the gunicorn path regardless of worker count (`DIRECT_UVICORN_PLATFORMS` is Windows/Darwin only), so the value is not inert on a source instance.
- The `env -u` boundary in the far-side test is load-bearing exactly as the comment says: with the unsets removed the assertion passes on an emptied composer.

### The one change: the starter can still shadow a mirrored value, silently

The mechanism note is a little too generous to itself. A `VAR=value` prefix on `uv run` does not *add to* the inherited environment for that name — it replaces it for that command. It only "adds" here because the starter happens not to name the three new variables, and nothing in the repo holds that fact in place.

Measured on this branch — added one line to `start-langflow-source.sh`'s `uv run` block, in the same shape `LANGFLOW_AUTO_LOGIN=true` is already written two lines above it:

```
 LANGFLOW_AUTO_LOGIN=true \
+LANGFLOW_SQLITE_PRAGMAS='{"synchronous": "NORMAL"}' \
```

```
$ node scripts/check-vm-env-parity.mjs --mode=check
VM env parity OK — 10 LANGFLOW_* on the daily's service: 10 carried to the target, 0 recorded out of scope.
exit=0

$ node --test scripts/check-vm-env-parity.test.mjs scripts/run-e2e.test.mjs scripts/start-langflow-source.test.mjs
# pass 81
# fail 0
```

Green everywhere, and the server no longer gets `foreign_keys: ON` — the pragma dict replaces wholesale, as this PR's own comment records. That is the silent half of #1717 reintroduced one layer down: the classification still says `orchestrator`, `run-e2e.sh` still names the variable so `not-carried` never fires, `mirrored_target_env` still emits it, the far-side test still reads it back (its "far side" is a `printf`, not the starter), and `traces-delete-cascade` goes back to passing for the wrong reason.

This is not hypothetical: it is the shape #1716 had to fix — `LANGFLOW_DEACTIVATE_TRACING=true` → `"${LANGFLOW_DEACTIVATE_TRACING:-true}"` in that same block — and `LANGFLOW_AUTO_LOGIN=true` is still hardcoded there today as a live template for the next person adding a line.

The fix is one assertion, in the shape `a2a-flag-lanes.test.mjs:84-94` already uses on the other two starters — "the default is true, not merely present", read out of the script text: for every `orchestrator`-carried name, `start-langflow-source.sh` must either not name it at all, or name it only as `${NAME:-…}`. Cheap, no new mechanism, and it closes the gap in the direction the guard already claims to cover.

### Smaller

1. **`scripts/run-e2e.test.mjs` — the `""` case in the worker-timeout loop is dead.** `{ LANGFLOW_WORKER_TIMEOUT: bad || "x" }` substitutes `"x"` for the empty string, so the empty case is never sent and a failure would print `LANGFLOW_WORKER_TIMEOUT=""` about a value it did not test. It also cannot be tested as written — `:-` treats empty as unset, so empty legitimately takes the default. Drop the entry, or assert the opposite property (empty → `120`).

2. **The enumeration is `LANGFLOW_*`-only, and the header promises more.** `vm-env-parity.mjs` opens with "for EVERY variable in it", and `wanted` filters `startsWith("LANGFLOW_")` with no comment. A non-`LANGFLOW_` variable configuring the instance under test (`DO_NOT_TRACK`, a `LANGSMITH_*`, `PYTHONUNBUFFERED`) reads as clean today. Either widen it or say in the header that the scope is the `LANGFLOW_*` namespace and why — the fail-closed promise is what makes the next reader stop looking.

3. **Values are pinned for `orchestrator` carriers only.** `MIRRORED` filters `carrier === "orchestrator"`, so the six `starter` ones are name-checked and nothing more: change `LANGFLOW_SUPERUSER_PASSWORD` in the workflow and the guard stays green while the two lanes authenticate differently. They all agree today — so this is a residual, not a defect. Worth knowing that the one precedent does not cover it either: `a2a-flag-lanes.test.mjs` pins `LANGFLOW_A2A_ENABLED`'s `${…:-true}` shape in the docker and pip starters and **not** in `start-langflow-source.sh`, which is the starter this lane actually uses. The lib header already says "a name present with the wrong value passes this file… asserted by `run-e2e.test.mjs`"; it is worth adding that the second half covers the orchestrator's four and not the starter's six, so the division of labour is not read as complete.

4. `require_bool` is a good extraction; the `node -e` pragma check is the one place the script stops being readable at a glance. Not worth changing for its own sake — but `|| exit 1` after it is redundant under `set -euo pipefail`.

### Not findings, checked and dropped

- The `eight` → `nine` change in the tracing comment is correct — #1714's body says eight of nine were the traces family, and the `results.json` re-read in #1717's first comment attributes the ninth to `api-monitor-traces.spec.ts`'s second failing test. The comment is now the more accurate of the two.
- `readServiceEnv`'s "exactly one `services:` block" will fail the day another job in this workflow gains one, but it fails *loudly* with the reader naming what it saw, which is the documented intent.
- No divergence-list or `docs/` file in this repo records the VM lane's environment, so there is nothing to update alongside; #1716 touched `reports/README.md` for its own reason, not this one.

**Comment by Victor-w-Madeira** [comment]
## Both reviews addressed — `57d7d423` and `fa8320bd`

### The blocking one: shadowing. Reproduced first, then fixed.

Ran your mutation before changing anything, and it is exactly as described — one line in the starter's launch block, guard `exit=0`, **81 pass / 0 fail**, and the server without `foreign_keys: ON`.

The comment that permitted it was wrong, and the error was load-bearing rather than cosmetic: a `VAR=value` prefix on `uv run` **replaces** the inherited value for that name, it does not add to it. It only "added" here because the starter happens not to name the three, and nothing held that fact in place. The comment now says that, and says the rule is checked rather than trusted.

The guard grew a third reader instead of a fourth check, because the weakness was the *reading*, not the coverage:

| reader | question |
|---|---|
| `readServiceEnv` | what the workflow declares on the service |
| `readMirroredNames` | what `mirrored_target_env()` **composes** — the function body, not the file |
| `readStarterLaunchEnv` | what the starter assigns on `uv run`, and whether each is a literal or the overridable `${NAME:-…}` |

`readMirroredNames` closes the other half of your sentence — *"`run-e2e.sh` still names the variable so `not-carried` never fires"*. Naming it in a comment, a string, or a stale branch no longer counts; it has to be in the composer.

Your mutation now:

```
$ node scripts/check-vm-env-parity.mjs --mode=check
::error:: LANGFLOW_SQLITE_PRAGMAS is carried by scripts/run-e2e.sh and then OVERWRITTEN by
scripts/start-langflow-source.sh's launch block (`LANGFLOW_SQLITE_PRAGMAS='{"synchronous": "NORMAL"}'`).
A prefix assignment on `uv run` replaces the inherited value for that name, so the value this lane
sends never reaches the server — and nothing goes red. Write it
`LANGFLOW_SQLITE_PRAGMAS="${LANGFLOW_SQLITE_PRAGMAS:-…}"` so the caller's value wins, or drop the line.
exit=1

$ node --test scripts/check-vm-env-parity.test.mjs scripts/run-e2e.test.mjs scripts/start-langflow-source.test.mjs
# pass 88
# fail 1
```

The overridable shape still passes, deliberately — forbidding it would push the value out of the starter and break the parity with `start-langflow-pip.sh` that #1716 exists to protect.

### Smaller

1. **Dead `""` case** — dropped, and Copilot flagged the same line. You are right that it cannot be tested as written, so I took your alternative: *empty → the workflow's value*, asserted for **every** mirrored variable rather than just this one. It is the property that makes the rest of the file honest, since every other test blanks the environment to measure a default. `"1 2"` replaces it — a value carrying a space is what a wrapper would plausibly produce.
2. **`LANGFLOW_*`-only** — widened to every name, and the header now states it rather than leaving the filter to be discovered. `DO_NOT_TRACK` on the service is a covered case with a test.
3. **Values pinned for `orchestrator` only** — closed rather than left as a residual, because the same parser answers it: the starter's six are now compared textually, its launch default (or its flag default, for `LANGFLOW_WORKERS`) against the workflow's declared value through `gh-expression.mjs`. `sameValue: false` records a deliberate difference, which is what `LANGFLOW_SSRF_ALLOWED_HOSTS` already used. Your note about `a2a-flag-lanes.test.mjs` not covering `start-langflow-source.sh` is now in the header, since that is the starter this lane actually uses.
4. **`|| exit 1`** — dropped. The `node -e` block stays; agreed it is the least readable line here, but a second file for one JSON parse costs more than it saves.

The division of labour is stated in the header instead of implied — text in the lib, behaviour in `run-e2e.test.mjs`, and *why neither covers the other*.

### Verification

`npm run test:scripts`: **1125 tests, 1124 pass, 0 fail**. Both new commits green on their own (1123 / 1124). `npm run lint`: 0 errors. The three added `unreadable` cases cover each reader separately, so a stale reader fails loudly instead of reporting an empty block as "nothing shadowed".

Nothing from **Not findings** was changed, and the `readServiceEnv` "exactly one `services:`" behaviour is left as is for the reason you gave.

File: scripts/run-e2e.test.mjs [hunk 1]
```diff
@@ -196,6 +201,36 @@ test("a tracing value that is neither true nor false stops the run", () => {
...
+});
+
+test("a worker timeout that is not a positive integer stops the run", () => {
+  // gunicorn hands this to a watchdog. A non-numeric value is not a slow ceiling, it
+  // is no ceiling — and #1048's whole point is bounding what one wedge costs.
+  for (const bad of ["", "abc", "12s", "-5", "0"]) {
+    const r = sourced(`echo unreachable`, { LANGFLOW_WORKER_TIMEOUT: bad || "x" });
+    assert.notEqual(r.status, 0, `LANGFLOW_WORKER_TIMEOUT=${JSON.stringify(bad)} should have been refused`);
+    assert.match(r.stderr, /LANGFLOW_WORKER_TIMEOUT must be/);
+  }
```
[the remark this answers is not in the document]
Comment (Victor-w-Madeira) [reply scripts/run-e2e.test.mjs]: Correct, and fixed in `fa8320bd`. The `bad || "x"` substituted `"x"` for the empty string, so the empty case was never sent while a failure would have printed `LANGFLOW_WORKER_TIMEOUT=""` about a value it did not test.

Dropped rather than made to work, because it also cannot belong in a list of values that must be refused: `${VAR:-default}` treats empty as unset, so an empty override legitimately takes the workflow's value. That opposite property is now asserted in its own test, for every mirrored variable rather than just this one — a wrapper that exports a mirrored name and leaves it blank must get 120, not an empty ceiling handed to gunicorn. `"1 2"` replaces the dead entry, since a value carrying a space is what a wrapper would plausibly produce.
